src/thumbnailer.cc

```diff
@@ -79,25 +79,68 @@ Thumbnailer::Status Thumbnailer::GetPictureStats(int ind, int* const pic_size,
     return kOk;
   }
 
-  WebPPicture new_pic;
-  if (!WebPPictureCopy(&frames_[ind].pic, &new_pic)) {
-    WebPPictureFree(&new_pic);
+  WebPPicture original_pic;
+  WebPPicture encoded_pic;
+  WebPMemoryWriter memory_writer;
+  WebPMemoryWriterInit(&memory_writer);
+
+  if (!WebPPictureCopy(&frames_[ind].pic, &encoded_pic)) {
+    WebPPictureFree(&encoded_pic);
     return kStatsError;
   }
 
+  // Lossy will modify the 'encoded_pic' but not lossless and near-lossless.
+  // Therefore, keep the encoded bitstream in the memory and decode it to
+  // compute PSNR correctly for near-lossless.
+  if (frames_[ind].config.lossless) {
+    encoded_pic.writer = WebPMemoryWrite;
+    encoded_pic.custom_ptr = (void*)&memory_writer;
+  } else {
+    if (!WebPPictureCopy(&frames_[ind].pic, &original_pic)) {
+      WebPPictureFree(&original_pic);
+      return kStatsError;
+    }
+  }
+
   WebPAuxStats stats;
-  new_pic.stats = &stats;
-  if (!WebPEncode(&frames_[ind].config, &new_pic)) {
-    WebPPictureFree(&new_pic);
+  encoded_pic.stats = &stats;
+  if (!WebPEncode(&frames_[ind].config, &encoded_pic)) {
+    WebPPictureFree(&encoded_pic);
     return kStatsError;
   }
 
-  *pic_size = new_pic.stats->coded_size;
+  if (frames_[ind].config.lossless) {
+    if (frames_[ind].config.near_lossless == 100) {
+      // Pure lossless: image was not modified, make 'original_pic' a view
```

Comment:
we could even skip computing the distortion in this case as it should be 99, but maybe not super important